### PR TITLE
MTDSA-19043 additional date validations tests

### PR DESCRIPTION
**Pull request by molafisoye**

### CODE REVIEW

File: test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala [hunk 1]
```diff
@@ -460,6 +460,64 @@ class CreateForeignPropertyPeriodSummaryValidatorSpec extends UnitSpec with Mock
...
+                |            "residentialFinancialCost":879.28,
+                |            "broughtFwdResidentialFinancialCost":846.13,
+                |            "other":138.92
+                |         }
+                |      }
+                |   ]
+                |}
+                |""".stripMargin
+            )
+          )) shouldBe List(FromDateOutOfRangeError)
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala:518]: Just a stylistic/convention thing, but could the result be assigned to a "val result"?  Then the final line is "result shouldBe ..."

File: test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala [hunk 2]
```diff
@@ -518,6 +576,66 @@ class CreateForeignPropertyPeriodSummaryValidatorSpec extends UnitSpec with Mock
...
+        val validator = setUpValidator()
+        validator.validate(
+          CreateForeignPropertyPeriodSummaryRawData(
+            validNino,
+            validBusinessId,
+            validTaxYear,
+            Json.parse(
+              """
+                |{
+                |   "fromDate":"2019-01-01",
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala:590]: As the json appears to be largely the same in each of these tests, perhaps they could be fetched from a "private def" (stuffed down at the end of the test class) and parameterised with the fromDate & toDate?

File: test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala (hunk 1 above)
Comment (basotia) [reply test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala:518]: @molafisoye I noticed the json used for test is used in multiple places, and only thing that is changing is mainly date or  some other attribute. Can we not extract common json and just replace the attribute we need to modify. I know this is not related to this test, but just by looking at Json it doesn't clearly tell me what all attributes are changing.

File: test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala (hunk 2 above)
Comment (basotia) [reply test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala]: @molafisoye My comment above is also same

**Review by basotia — DISMISSED** [review]
looks good

File: test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala [hunk 3]
```diff
@@ -2467,10 +2115,62 @@ class CreateForeignPropertyPeriodSummaryValidatorSpec extends UnitSpec with Mock
...
         val validator = setUpValidator()
-        validator.validate(CreateForeignPropertyPeriodSummaryRawData("A12344A", "20178", "232301", requestBodyJson)) shouldBe
+        val result    = validator.validate(CreateForeignPropertyPeriodSummaryRawData("A12344A", "20178", "232301", requestBodyJson))
+        result shouldBe
           List(NinoFormatError, TaxYearFormatError, BusinessIdFormatError)
       }
     }
   }
 
+  trait TestData {
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateForeignPropertyPeriodSummaryValidatorSpec.scala:2125]: This doesn't need to be in its own trait; nothing's really gained, but it does have some drawbacks, mainly:
1. turns the test into an instance of a class, so you have to deal with private/public vals, specifying types etc
2. "requestBody" ends up with multiple instance creations (one per TestData subclass) for no particular reason
3. it's just added complexity
Easier to just put requestBody as a private val.  Better would be to make it a private def, and parameterise it for the fromDate & toDate.

File: test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala [hunk 4]
```diff
@@ -327,6 +327,17 @@ class CreateUkPropertyPeriodSummaryValidatorSpec extends UnitSpec with JsonError
           )) shouldBe List(FromDateFormatError)
       }
 
+      "the fromDate is out of range" in {
+        val validator = setUpValidator()
+        validator.validate(
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala:332]: can you assign this to a result, so it's "result shouldBe ..."

File: test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala [hunk 5]
```diff
@@ -327,6 +327,17 @@ class CreateUkPropertyPeriodSummaryValidatorSpec extends UnitSpec with JsonError
...
+          CreateUkPropertyPeriodSummaryRawData(
+            validNino,
+            taxYear,
+            validBusinessId,
+            requestBodyJson.update("fromDate", JsString("1569-10-01"))
+          )) shouldBe List(FromDateOutOfRangeError)
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala:338]: same

File: test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala [hunk 6]
```diff
@@ -338,6 +349,17 @@ class CreateUkPropertyPeriodSummaryValidatorSpec extends UnitSpec with JsonError
...
 
+      "the toDate is out of range" in {
+        val validator = setUpValidator()
+        validator.validate(
+          CreateUkPropertyPeriodSummaryRawData(
+            validNino,
+            taxYear,
+            validBusinessId,
+            requestBodyJson.update("toDate", JsString("3490-10-01"))
+          )) shouldBe List(ToDateOutOfRangeError)
```
Comment (mattstephens-hmrc) [inline test/v2/controllers/requestParsers/validators/CreateUkPropertyPeriodSummaryValidatorSpec.scala:360]: same

**Review by mattstephens-hmrc — APPROVED** [review]

**Review by mattstephens-hmrc — APPROVED** [review]

**Pull request merged**